File: 2020-10-30-client-libraries/go/demo_requests.go
package main

import (
  "fmt"

  "github.com/stripe/stripe-go"
  "github.com/stripe/stripe-go/customer"
  // "github.com/stripe/stripe-go/paymentintent"
  // "github.com/stripe/stripe-go/invoice"
)

func main() {
  stripe.Key = "sk_test_51EceeUCZ6qsJgndJDSi5feJtMJs4e4SOOQL7TIGtQyIA7GsyJczBvxvrFsuB71OkREXySaFDzcjLYb2IoDmuX1jL00e4sdsH5H"
  fmt.Println("Making requests!")

  // Create a customer with no params
  // c, _ := customer.New(nil)
  // fmt.Println(c)

  // Fetch a customer
  // c, _ := customer.Get("cus_IVTjl1Y4JUVONJ", nil)
  // fmt.Println(c)

  // Create a customer with scalar values.
  // params := &stripe.CustomerParams{
  //   Email: stripe.String("jenny.rosen@example.com"),
  //   Name: stripe.String("Jenny Rosen"),
  // }
  // // email=jenny.rosen@example.com&name=Jenny%20Rosen
  // c, _ := customer.New(params)
  // fmt.Println(c)
  // fmt.Println(c.Name)
  // fmt.Println(c.Email)

  // Create a customer with an enum value
  // params := &stripe.CustomerParams{
  //   TaxExempt: stripe.String(string(stripe.CustomerTaxExemptReverse)),
  // }
  // c, _ := customer.New(params)
  // fmt.Println(c.TaxExempt)

  // Create a customer with nested params
  // params := &stripe.CustomerParams{
  //   PaymentMethod: stripe.String("pm_card_visa"),
  //   InvoiceSettings: &stripe.CustomerInvoiceSettingsParams{
  //     DefaultPaymentMethod: stripe.String("pm_card_visa"),
  //   },
  // }
  // c, _ := customer.New(params)
  // fmt.Println(c.InvoiceSettings.DefaultPaymentMethod)

  // Create a customer with a list of Strings
  // params := &stripe.CustomerParams{
  //   PreferredLocales: stripe.StringSlice(
  //     []string{"en", "es"},
  //   ),
  // }
  // c, _ := customer.New(params)
  // fmt.Println(c.PreferredLocales)

  // Update the email address for a customer
  // params := &stripe.CustomerParams{
  //   Email: stripe.String("jr-2@example.com"),
  // }
  // c, _ := customer.Update(
  //   "cus_IVTjl1Y4JUVONJ",
  //   params,
  // )
  // fmt.Println(c.ID)
  // fmt.Println(c.Email)

  // Update customer with nested list of objects
  // params := &stripe.CustomerParams{
  //   InvoiceSettings: &stripe.CustomerInvoiceSettingsParams{
  //     CustomFields: []*stripe.CustomerInvoiceCustomFieldParams{
  //       {
  //         Name: stripe.String("VAT"),
  //         Value: stripe.String("123ABC"),
  //       },
  //     },
  //   },
  // }
  // c, _ := customer.Update(
  //   "cus_IVTjl1Y4JUVONJ",
  //   params,
  // )
  // fmt.Println(*c.InvoiceSettings.CustomFields[0].Name)
  // fmt.Println(*c.InvoiceSettings.CustomFields[0].Value)

  // Fetch list of customers
  // params := &stripe.CustomerListParams{}
  // params.Single = true
  // i := customer.List(params)
  // for i.Next() {
  //   c := i.Customer()
  //   fmt.Println(c.ID)
  // }

  // Filter the list of customers by email
  // params := &stripe.CustomerListParams{
  //   Email: stripe.String("jenny.rosen@example.com"),
  // }
  // // /v1/customers?email=jenny.rosen@example.com
  // params.Single = true
  // i := customer.List(params)
  // for i.Next() {
  //   c := i.Customer()
  //   fmt.Println(c.ID)
  //   fmt.Println(c.Email)
  // }

  // Delete a customer
  // c, _ := customer.Del("cus_IVTjl1Y4JUVONJ", nil)
  // fmt.Println(c)

  // First, create a payment intent to confirm later
  // params := &stripe.PaymentIntentParams{
  //   Amount: stripe.Int64(1000),
  //   Currency: stripe.String(string(stripe.CurrencyUSD)),
  // }
  // pi, _ := paymentintent.New(params)
  // fmt.Println(pi.ID)
  // fmt.Println(pi.Status)

  // Second, confirm payment intent
  // params := &stripe.PaymentIntentConfirmParams{
  //   PaymentMethod: stripe.String("pm_card_visa"),
  // }
  // pi, _ := paymentintent.Confirm(
  //   "pi_1HuTDeCZ6qsJgndJ7QUDL2cL",
  //   params,
  // )
  // fmt.Println(pi.ID)
  // fmt.Println(pi.Status)
  // Nested service method for line items on invoices.
  // params := &stripe.InvoiceLineListParams{
  //   ID: stripe.String("in_1HuPYjCZ6qsJgndJhXQpNBXm"),
  // }
  // // /v1/invoices/in_xxx/lines
  // params.Filters.AddFilter("limit", "", "5")
  // i := invoice.ListLines(params)
  // for i.Next() {
  //   line := i.InvoiceLine()
  //   fmt.Println(line.ID)
  // }

  // Create a customer on a connected account
  params := &stripe.CustomerParams{
    Email: stripe.String("jenny.rosen@example.com"),
  }
  params.SetStripeAccount("acct_1Ey3h1BqeQ4DKpna")

  c, _ := customer.New(params)
  fmt.Println(c.ID)
  fmt.Println(c.Email)
}

File: 2020-10-30-client-libraries/node/demo_requests.js
const stripe = require('stripe')('sk_test...');

(async () => {
  console.log('Making requests!');
  // Create a customer with no params.
  // const customer = await stripe.customers.create();
  // console.log(customer);

  // Fetch a customer
  // const cus = await stripe.customers.retrieve(
  //   'cus_IVRyRXG6HUVua2',
  // );
  // console.log(cus);

  // Craete customer with scalar values
  // const cus = await stripe.customers.create({
  //   email: 'jenny.rosen@example.com',
  //   name: 'Jenny Rosen',
  // })
  // // email=jenny.rosen@examp.com&name=Jenny%20Rosen
  // console.log(cus)

  // Create customer with enum value
  // try {
  //   const cus = await stripe.customers.create({
  //     tax_exempt: 'invalid'
  //   });
  //   console.log(cus.tax_exempt);
  // } catch (e) {
  //   console.log(e);
  // }

  // Make a request with invalid params
  // try {
  //   const cus = await stripe.customers.create({
  //     email_address: 'jennyrosen@example.com'
  //   });
  //   console.log(cus.email);
  // } catch (e) {
  //   console.log(e);
  // }

  // Create a customer with nested object
  // const cus = await stripe.customers.create({
  //   payment_method: 'pm_card_visa',
  //   invoice_settings: {
  //     default_payment_method: 'pm_card_visa',
  //   }
  // });
  // console.log(cus);

  // Create a customer with a list of strings.
  // const cus = await stripe.customers.create({
  //   preferred_locales: ['en', 'es']
  // });
  // console.log(cus);

  // Update the email address for a customer
  // const cus = await stripe.customers.update(
  //   'cus_IVRyRXG6HUVua2', {
  //     email: 'jr2@example.com',
  //   }
  // );
  // console.log(cus.id);
  // console.log(cus.email);

  // Update a customer with nested object data
  // const cus = await stripe.customers.update(
  //   'cus_IVRyRXG6HUVua2', {
  //     invoice_settings: {
  //       custom_fields: [{
  //         name: 'VAT',
  //         value: '123ABC'
  //       }]
  //     }
  //   }
  // )
  // console.log(cus.invoice_settings);

  // Fetch a list of customers
  // const customers = await stripe.customers.list();
  // console.log(customers);
  // console.log(customers.data.map(c => c.id))

  // Filter a list of customers by email
  // const customers = await stripe.customers.list({
  //   email: 'jenny.rosen@example.com'
  // });
  // // /v1/customers?email=jenny.rosen@example.com
  // console.log(customers.data.map(c => c.id))
  // console.log(customers.data.map(c => c.email))

  // Deleting an object
  // const cus = await stripe.customers.del(
  //   'cus_IVRyRXG6HUVua2'
  // )
  // console.log(cus);
  //
  // Custom methods
  // First create a payment intent to confirm.
  // const intent = await stripe.paymentIntents.create({
  //   amount: 1000,
  //   currency: 'usd'
  // });
  // console.log(intent.id);
  // console.log(intent.status);

  // Second, confirm the payment intent
  // const intent = await stripe
  //   .paymentIntents
  //   .confirm(
  //     'pi_1HuSWVCZ6qsJgndJr2xWV3v8', {
  //       payment_method: 'pm_card_visa',
  //     }
  //   )
  // console.log(intent.id);
  // console.log(intent.status);

  // Fetch lines from invoice with nested service
  // method.
  // const lines = await stripe
  //   .invoices
  //   .listLineItems(
  //     'in_1HuPYjCZ6qsJgndJhXQpNBXm', {
  //       limit: 5
  //     }
  //   );
  // console.log(lines);

  // Pass a request header.
  const cus = await stripe.customers.create({
    email: 'jenny.rosen@example.com',
  }, {
    stripeAccount: 'acct_1Ey3h1BqeQ4DKpna'
  })
  console.log(cus);

})();
